### DISCUSSION TITLE
Help with ProtonSDK's Surface and SurfaceAnim components

### DISCUSSION THREAD

**Discussion by rainerhagert** [SethRobinson/proton#32] [Q&A] [answered]
Hey,

I am working on a 2D platformer game inspired by Growtopia, using Proton SDK. I'm reaching out to ask for a bit of help with rendering the player character.

In the initial stages of development, my focus is on rendering the basic player character without diving into animations just yet. 

However, I've encountered some challenges in understanding how to use the "Surface" and "SurfaceAnim" components. Unfortunately, the available documentation is limited, making this task more complex than anticipated.

I've already grasped the basics of Proton SDK to some extent, but I'm struggling to properly draw the player along with its animations.

I kindly request any additional information, guidance, code snippets, or best practices you can offer to assist me in rendering the player character. Your insights at this point would be a valuable addition to my project, helping me establish a strong foundation before I move on to more advanced animations.

Thank you for your time and consideration.

**Comment by SethRobinson** [comment, marked as answer]
Sure,

Hmm, you're right, there is really no documentation on this.  A surface loads a graphics, and allows you to do 2d blits to the screen easily.

SurfaceAnim subclasses Surface and adds a few extra things for animations and blitting, so you pretty much want to always use that.

The RTBareBones example shows a Surface being loaded and blitted - a lot of the code is commented out, but it shows valid examples of different ways to "blit" it to the screen.

If you look through the SurfaceAnim header, you can see a lot of different Blit functions, some let you tint the image with colors, or rotate it around a specific point.  

```
//you need to load a graphic.  Depending on what Proton has enabled, it can be an .rttex, bmp, jpg, png etc.
SurfaceAnim m_surf;  //uh, normally you wouldn't load this every frame... this is just a dumb example

	if (!m_surf.IsLoaded())
	{
		m_surf.LoadFile("interface/test.rttex");
	}

	//blit the logo with the Y mirrored
	//rtRect texRect = rtRect(0, m_surf.GetHeight(), m_surf.GetWidth(), 0);
	//rtRect destRect = rtRect(0,0, m_surf.GetWidth(), m_surf.GetHeight());
	//m_surf.BlitEx(destRect, texRect);

	//make the logo spin like a wheel, whee!
	//m_surf.BlitEx(destRect, texRect, MAKE_RGBA(255,255,255,255) , 180*SinGamePulseByMS(3000), CL_Vec2f(m_surf.GetWidth()/2,m_surf.GetHeight()/2));

	//blit it normally
	m_surf.Blit(0, 0);
	//m_surf.Blit(100, 100);

	m_surf.BlitScaled(100, 200, CL_Vec2f(1,1), ALIGNMENT_CENTER, MAKE_RGBA(255,255,255,255), SinGamePulseByMS(3000)*360);

	m_surf.BlitRotated(400, 200, CL_Vec2f(0.2f,0.2f), ALIGNMENT_CENTER, MAKE_RGBA(255,255,255,255), SinGamePulseByMS(4000)*360,
		CL_Vec2f(20,-20), NULL);
```

You can also say "this giant bmp is actually a grid of frames, so treat it like an animation" -  just call SetupAnim or SetupAnimBySize after loading it.

Then by using BlitAnim, BlitScaledAnim, or BlitRotatedAnim you can also specify a certain frame to draw.

You might want to check out the Loony Ladders game example, I think it uses SurfaceAnim for everything.

**Reply by rainerhagert** [reply]
Thanks a lot for the response! I used your code and managed to draw a basic player in my game. I'll try to set up animations later on, but I think I understood how they work. I'll update this thread if I run into issues with them.

**Reply by SethRobinson** [reply]
Just a quick note, I actually added a bug that made SurfaceAnim draw incorrectly when animations were used two days ago, it's now fixed. 

Just hoping you didn't get hit with that... would make learning quite confusing to say the least. :)  That's what I get for letting ChatGPT add stuff without properly testing it.

**Reply by rainerhagert** [reply]
Great to hear that the issue has been fixed! I'm using an older version of Proton, so luckily, I didn't encounter that bug. Thanks for the heads up, and don't worry about it - bugs happen to the best of us!
